### DISCUSSION TITLE
Select all my branches with the goal of rebasing them all on the trunk

### DISCUSSION THREAD

**Discussion by jennings** [jj-vcs/jj#4974] [Nice Revsets] [answered]
After fetching the latest `main`, I want to rebase all of my work onto it.

**Comment by jennings** [comment]
The revset `mutable() & mine()` may be enough to select all branches you're interested in. Use `--branch` to rebase all branches containing those commits (short form: `-b`).

```shell
$ jj rebase -b 'all:mutable() & mine()' -d trunk()
```

This revset was [written by pelme in Discord](https://discord.com/channels/968932220549103686/968932220549103689/1310674303192141844).

**Comment by jennings** [comment, marked as answer]
After thinking about this problem more, I think the safer version of this is:

```shell
$ jj rebase -d main -s 'roots(trunk()..) & mutable()'
```

---

At first, it might seem like simply rebasing `mutable()` would work:

```shell
# WARNING: May cause problems, see below
$ jj rebase -d trunk() -r mutable()
```

However, that can accidentally transplant revisions. Suppose you pushed two revisions, and someone else put a bookmark on the first:

```
◆    M (main)
|
| ○  C (my-branch)
| |
| ◆  B (someone-else@origin)
|/
◆    A
```

jj's default settings would make revision B immutable because an untracked remote bookmark points to it. If you rebase `-r mutable()`, revision C will be transplanted:

```
○    C' (my-branch)
|
◆    M (main)
|
| ◆  B (someone-else@origin)
|/
◆
```

So, you probably want to rebase all mutable changes whose parent is in the main branch. The revset below will select the base of each branch off `main`, as long as it's mutable. We use `-s` instead of `-r` so we only have to select the base of the branch, not the entire thing.

```shell
$ jj rebase -d main -s 'roots(trunk()..) & mutable()'
```

Explanation:

```
          trunk()..                    All commits not in the trunk
    roots(trunk()..)                   Take only the root commits of those...
    roots(trunk()..) & mutable()       ...as long as they're mutable
```

This does lose the `mine()` filter. You could update the revset to `roots(trunk()..) & mutable() & mine()`, though that filter will only apply to the first revision in the branch.

**Reply by scott2000** [reply]
Instead of `(::trunk())+ & mutable()`, using `roots(trunk()..) & mutable()` might be more efficient? I believe `(::trunk())+` will be expensive in repos with a large number of commits in `::trunk()`.

**Reply by jennings** [reply]
Good to know! Is this the reasoning?

- `(::trunk())+` contains all the commits in `..trunk()` in addition to the roots that we care about
- So, the left side of the intersection has high cardinality
- On the other hand, `roots(trunk()..)` has already eliminated the high cardinality in typical repos?

**Reply by jennings** [reply]
Ah, in fact, `roots(trunk()..) & mutable` is more correct, too. In this scenario, I think `(::trunk())+` selects `M` and rips it off of `A`, whereas `roots(trunk()..)` doesn't:

```
M
|\
A |  (mutable)
| |
◆ /  trunk()
|/
◆
```

If you have unrelated histories in the same repo, then I guess you'd need to use `(..trunk())+`, but that's probably not too common.

**Reply by scott2000** [reply]
It's more to do with the fact that both `x+` and `roots(x)` currently require eagerly evaluating `x` in its entirety. I was just guessing that `::trunk()` is likely to be large since it contains the entire history of the repo, so it would require more work to process it.

But I tested it in some large repos just now, and it seems like the difference is pretty small actually (although I didn't test very thoroughly). The commit index is so fast that it doesn't really matter that much. So maybe nevermind!
